### PR TITLE
Added option to stop emulators. (#15)

### PR DESCRIPTION
**Pull request by yunikkk**

### CODE REVIEW

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt [hunk 1]
```diff
@@ -27,61 +28,81 @@ private class JCommanderArgs {
...
 
 fun parseArgs(rawArgs: List<String>): List<Args> {
-    if (PARAMETER_HELP_NAMES.firstOrNull { rawArgs.contains(it) } != null) {
+    if (COMMAND_HELP_NAMES.firstOrNull { rawArgs.contains(it) } != null) {
         JCommander(JCommanderArgs()).usage()
         System.exit(0)
     }
 
+    if (COMMAND_STOP_NAMES.firstOrNull { rawArgs.contains(it) } != null) {
+        stopAllEmulators().toBlocking().firstOrDefault(Unit)
```
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt]: Oh, let's move it to main(), parseArgs() should definitely not be doing anything like that

Maybe return sealed class of parse result?

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt [hunk 2]
```diff
@@ -205,3 +204,25 @@ fun Long.nanosAsSeconds(): Float = NANOSECONDS.toMillis(this) / 1000f
...
+
+fun stopAllEmulators(): Observable<Unit> =
+        connectedEmulators()
+                .switchMap { emulators ->
+                    Observable.zip(
+                            emulators
+                                    .map { emulator ->
+                                        process(
+                                                listOf(adb, "-s", emulator.id, "emu", "kill"),
+                                                timeout = 5 to SECONDS
```
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt]: Hm, I'm pretty sure we'll overpass this timeout on our highloaded CI

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt [hunk 3]
```diff
@@ -205,3 +204,25 @@ fun Long.nanosAsSeconds(): Float = NANOSECONDS.toMillis(this) / 1000f
 
 private fun outputFileForEmulator(args: Args) = File("${args.emulatorName}.output")
         .apply { deleteOnExit() }
+
+private fun connectedEmulators(): Observable<Iterable<AdbDevice>> =
+        connectedAdbDevices().map { it.filter { it.isEmulator } }
+
+fun stopAllEmulators(): Observable<Unit> =
```
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt]: Maybe Completable?

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt (hunk 2 above)
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt]: I guess something like 15 seconds would be a better default?

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt (hunk 1 above)
Comment (yunikkk) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt]: Makes sense, thought about it. Decided to go easy way, though=\

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt (hunk 3 above)
Comment (yunikkk) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt]: Ok

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt (hunk 2 above)
Comment (yunikkk) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Emulators.kt]: Ok, lets try 15

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt (hunk 1 above)
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt]: Hah, it'd never easy man, never easy

File: dependencies.gradle [hunk 4]
```diff
@@ -2,7 +2,7 @@ ext.versions = [
         kotlin       : '1.1.1',
 
         rxJava       : '1.2.9',
-        jCommander   : '1.66',
+        jCommander   : '1.71',
```
Comment (yunikkk) [inline dependencies.gradle]: Old version was hiding JCommander.newBuilder

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt [hunk 5]
```diff
@@ -2,140 +2,143 @@ package com.gojuno.swarmer
...
-                if (indexOfFirst == 0 && size > 1) {
-                    subList(indexOfFirst + 1, size).indexOfFirstParameterEmulatorName().let { if (it == -1) it else it + 1 }
-                } else {
-                    -1
+fun parseCommand(rawArgs: List<String>) = Commands.fromStringAlias(rawArgs[0])
+
+fun parseStartArguments(rawArgs: List<String>): List<Commands.Start> =
+        rawArgs
+                .subList(1, rawArgs.size) // skip command
+                .fold(ArrayList<ArrayList<String>>()) { accumulator, value ->
```
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Args.kt]: Oh finally

File: swarmer/src/main/kotlin/com/gojuno/swarmer/Main.kt [hunk 6]
```diff
@@ -1,35 +1,23 @@
...
-    val startedEmulators = connectedAdbDevices()
-            .doOnNext { log("Already running emulators: $it") }
-            .flatMap {
-                val startEmulators: List<Observable<Emulator>> = args
-                        .map { startEmulator(it, availablePortsSemaphore) }
-                        .map { it.subscribeOn(io()) } // So each emulator will start in parallel.
-                        .map { it.doOnNext { log("Emulator $it is ready.") } }
-
-                Observable.zip(startEmulators, { startedEmulator -> startedEmulator })
+    rawArgs.toList().apply {
```
Comment (artem-zinnatullin) [inline swarmer/src/main/kotlin/com/gojuno/swarmer/Main.kt]: this is awesome!

File: swarmer/src/test/kotlin/com/gojuno/swarmer/ArgsSpec.kt [hunk 7]
```diff
@@ -2,39 +2,55 @@ package com.gojuno.swarmer
...
-                    pathToConfigIni = "test_path_to_config_ini",
-                    emulatorStartOptions = emptyList(),
-                    emulatorStartTimeoutSeconds = 180,
-                    redirectLogcatTo = null
+                    pathToConfigIni = "test_path_to_config_ini"
             )))
         }
     }
 
+    fun SpecBody.onEmulatorNameVariants() {
```
Comment (artem-zinnatullin) [inline swarmer/src/test/kotlin/com/gojuno/swarmer/ArgsSpec.kt]: lol, we need to del this

File: swarmer/src/test/kotlin/com/gojuno/swarmer/EmulatorsSpec.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import com.nhaarman.mockito_kotlin.whenever
+import org.jetbrains.spek.api.Spek
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import org.mockito.Mockito
+import rx.Completable
+import rx.Single
+import java.util.concurrent.TimeUnit
+
+class EmulatorsSpec : Spek({
```
Comment (artem-zinnatullin) [inline swarmer/src/test/kotlin/com/gojuno/swarmer/EmulatorsSpec.kt]: 👍

File: swarmer/src/test/kotlin/com/gojuno/swarmer/ArgsSpec.kt (hunk 7 above)
Comment (yunikkk) [inline swarmer/src/test/kotlin/com/gojuno/swarmer/ArgsSpec.kt]: NO

**Pull request merged**